File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
"""
Time = O(n)
Space = O(n)
"""
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        dic = defaultdict(int)
        dic[0] = 1
        acc_sum = 0
        count = 0
        for i in range(len(nums)):
            acc_sum += nums[i]
            if (acc_sum - k) in dic:
                count += dic[acc_sum - k]
            dic[acc_sum] += 1
            
        return count
            
                

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        /* use hashmap to count the # of same value of sums in different subarray */
        unordered_map<int, int> counter;
        int res = 0, accSum = 0;
        counter[0] = 1;
        
        for (int num: nums) {
            accSum += num;
            if (counter.find(accSum - k) != counter.end()) {
                /* found current sequence can sum equals k by removing prefix */
                res += counter[accSum - k];
            }
            counter[accSum]++;
        }
        
        return res;
    }
};